src/adapter/src/catalog/migrate.rs

```diff
@@ -182,6 +210,576 @@ pub(crate) async fn migrate(
+                // already been migrated.
+                let tables_for_source =
+                    items_with_statements_copied
+                        .iter()
+                        .any(|(_, statement)| match statement {
+                            Statement::CreateTableFromSource(stmt) => {
+                                let source: CatalogItemId = match &stmt.source {
+                                    RawItemName::Name(_) => {
+                                        unreachable!("tables store source as ID")
+                                    }
+                                    RawItemName::Id(source_id, _, _) => {
+                                        source_id.parse().expect("valid id")
+                                    }
+                                };
+                                source == item.id
+                            }
+                            _ => false,
+                        });
+                if tables_for_source
+                    && envelope.is_none()
+                    && format.is_none()
+                    && include_metadata.is_empty()
+                {
+                    info!("migrate: skipping already migrated source: {}", name);
+                    continue;
+                }
+
+                // Use the current source name as the new table name, and rename the source to
+                // `<source_name>_source`. This is intended to allow users to continue using
+                // queries that reference the source name, since they will now need to query the
+                // table instead.
+
+                // First find an unused name within the same schema to avoid conflicts.
+                let mut new_source_item_name = format!("{}_source", item.name);
+                let mut new_source_name_inner =
+                    format!("{}_source", name.0.last().expect("at least one ident"));
+                let mut i = 0;
+                while item_names_per_schema
+                    .get(&item.schema_id)
+                    .expect("schema must exist")
+                    .contains(&new_source_item_name)
+                {
+                    new_source_item_name = format!("{}_source_{}", item.name, i);
+                    new_source_name_inner = format!(
+                        "{}_source_{}",
+                        name.0.last().expect("at least one ident"),
+                        i
+                    );
+                    i += 1;
+                }
+                // We will use the original item name for the new table item.
+                let table_item_name = item.name.clone();
+
+                // Update the source item/statement to use the new name.
+                let mut new_source_name = name.clone();
+                *new_source_name.0.last_mut().expect("at least one ident") =
+                    Ident::new_unchecked(new_source_name_inner);
```

Comment:
You should be able to replace all of this with [`Ident::try_generate_name`](https://github.com/MaterializeInc/materialize/blob/main/src/sql-parser/src/ast/defs/name.rs#L153), or if that doesn't work I would recommend at least using `Ident::append_lossy` to avoid the bug where the new `{}_source` name is longer than max size.